pkg/nmi/server/server.go

```diff
@@ -380,3 +386,29 @@ func handleTermination() {
 	log.Infof("Exiting with %v", exitCode)
 	os.Exit(exitCode)
 }
+
+func listPodIDsWithRetry(kubeClient k8s.Client, logger *log.Entry, podns, podname, rqClientID string, maxAttempts int) (*[]aadpodid.AzureIdentity, error) {
+	attempt := 0
+	var err error
+	var podIDs *[]aadpodid.AzureIdentity
+
+	for {
+		podIDs, err = kubeClient.ListPodIds(podns, podname)
+		if err == nil && len(*podIDs) != 0 {
+			if len(rqClientID) == 0 {
+				return podIDs, nil
+			}
+			// if client id exists in request, we need to ensure the identity with this client
+			// exists and is in Assigned state
+			for _, podID := range *podIDs {
+				if strings.EqualFold(rqClientID, podID.Spec.ClientID) {
+					return podIDs, nil
+				}
+			}
+		}
+		attempt++
+		logger.Warningf("failed to get assigned ids for pod:%s/%s, retrying attempt: %d", podns, podname, attempt)
+		time.Sleep(listPodIDsRetryIntervalInSeconds * time.Second)
+	}
+	return nil, fmt.Errorf("getting assigned identities for pod %s/%s failed after %d attempts. Error: %v", podns, podname, maxAttempts, err)
```

Comment:
will this error propagate to the caller in the application ?

Reply:
Yeah, this will be propagated if it fails after retries.